diff --git a/mm/ksm.c b/mm/ksm.c
index 6cb60f4..b9520e2 100644
--- a/mm/ksm.c
+++ b/mm/ksm.c
@@ -991,6 +991,28 @@ static u32 calc_checksum(struct page *page)
 	return checksum;
 }
 
+#ifdef CONFIG_X86_64
+int memcmpx86_64(void *s1, void *s2, size_t n)
+{
+	size_t num = n / 8;
+	register int res;
+
+	__asm__ __volatile__
+		(
+		 "testq %q3,%q3\n\t"
+		 "repe; cmpsq\n\t"
+		 "je        1f\n\t"
+		 "sbbq      %q0,%q0\n\t"
+		 "orq       $1,%q0\n"
+		 "1:"
+		 : "=&a" (res), "+&S" (s1), "+&D" (s2), "+&c" (num)
+		 : "0" (0)
+		 : "cc");
+
+	return res;
+}
+#endif
+
 static int memcmp_pages(struct page *page1, struct page *page2)
 {
 	char *addr1, *addr2;
@@ -998,7 +1020,11 @@ static int memcmp_pages(struct page *page1, struct page *page2)
 
 	addr1 = kmap_atomic(page1);
 	addr2 = kmap_atomic(page2);
+#ifdef CONFIG_X86_64
+	ret = memcmpx86_64(addr1, addr2, PAGE_SIZE);
+#else
 	ret = memcmp(addr1, addr2, PAGE_SIZE);
+#endif
 	kunmap_atomic(addr2);
 	kunmap_atomic(addr1);
 	return ret;
